### PR TITLE
G2p

### PR DESCRIPTION
**Pull request by dcolligan**
Merge of G2P into master

@bwalsh

### CODE REVIEW

File: doc/source/api/proposed_schema_changes.md [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+Many types rely heavily on the concept of an [OntologyTerm](https://github.com/ga4gh/schemas/blob/be171b00a5f164836dfd40ea5ae75ea56924d316/src/main/resources/avro/ontologies.avdl#L10)  (see end of document for discussion on usage of OntologyTerms).
+
+
+##Implementation
+
+![image](https://cloud.githubusercontent.com/assets/47808/15920813/d70195a0-2dd1-11e6-9c74-ba552735404d.png)
+
+
+### Source Code
+* [Front End](https://github.com/ohsu-computational-biology/server/blob/g2p-2.0/ga4gh/frontend.py) '/genotypes/search', '/phenotypes/search', '/genotypephenotypes/search'
```
Comment (sarahhunt) [inline doc/source/api/proposed_schema_changes.md]: Should these be pointing to the ga4gh repo now?

File: doc/source/api/proposed_schema_changes.md [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+* [Back End](https://github.com/ohsu-computational-biology/server/blob/g2p-2.0/ga4gh/backend.py) 'runSearchGenotypePhenotypes', 'runSearchPhenotypes', 'runSearchGenotypes'
+* [Datamodel](https://github.com/ohsu-computational-biology/server/blob/g2p-2.0/ga4gh/datamodel/genotype_phenotype.py) 'getAssociations'
+
+### Tests
+* [End to End](https://github.com/ohsu-computational-biology/server/blob/g2p-2.0/tests/end_to_end/test_g2p.py)
+
+   **Help Wanted:** Any or all use cases and scenarios
+
+
+### Acceptance
```
Comment (sarahhunt) [inline doc/source/api/proposed_schema_changes.md]: Not useful here.

File: doc/source/api/proposed_schema_changes.md [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+Create an PhenotypeQuery using description field.
+``{"description": "inflammatory bowel disease",...}``
+The system responds with Phenotypes that match on OntologyTerm.description
+Note that you can wildcard description.
+``{"description": ".*bowel.*",...}``
+[Supported regex](https://www.w3.org/TR/xpath-functions/#regex-syntax)
+
+___
+
+###``/genotypes/search``
```
Comment (sarahhunt) [inline doc/source/api/proposed_schema_changes.md]: This is no longer part of the API

File: doc/source/api/proposed_schema_changes.md [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+   phenotype.
+
+For specifics of the json representations, please see the
+`server <https://github.com/ga4gh/server>`__ and
+`compliance <https://github.com/ga4gh/compliance>`__ repositories.
+
+Ontologies
+----------
+
+**Usage:** Multiple ontology terms can be supplied e.g. to describe a series
```
Comment (sarahhunt) [inline doc/source/api/proposed_schema_changes.md]: refers to ontology.avdl which was replaced with metadata.proto

File: doc/source/api/proposed_schema_changes.md [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+
+**Definitions:**
+
+*Annotation* - A free text annotation which is not an
+ontology term describing some attribute. Annotations have associations
+with OntologyTerms to allow these to be added after annotations are
+captured. OntologyTerms are preferred over Annotations in all cases.
+Annotations can be used in conjucntion with OntologyTerms
+
+*OntologyTerm* - the preferred term for the class in question. For example
```
Comment (sarahhunt) [inline doc/source/api/proposed_schema_changes.md]: It would be better to refer to the metadata documentation here - this is already out of date.

File: doc/source/api/proposed_schema_changes.md [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+
+![image](https://cloud.githubusercontent.com/assets/47808/14397329/99fc7c30-fd91-11e5-8346-e95f97bfb78c.png)
+
+
+### Convenience endpoints
+
+(As needed)
+
+```
+GET /associationsets/<phenotypeAssociationSetId>/feature/<id>/associations
```
Comment (sarahhunt) [inline doc/source/api/proposed_schema_changes.md]: Are these part of the current spec?

File: doc/source/api/proposed_schema_changes.md [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+references to ontological concepts and/or controlled vocabularies within
+AVRO. The structures provided are not intended for de novo modeling of
+ontologies, or representing complete ontologies within AVRO. References
+to e.g. classes from external ontologies or controlled vocabularies
+should be interpreted only in their original context i.e. the source
+ontology.
+
+Due to the flexibility of the data model, users have a number of options
+for specifying each query term
+[feature](https://github.com/ga4gh/schemas/blob/be171b00a5f164836dfd40ea5ae75ea56924d316/src/main/resources/avro/genotypephenotypemethods.avdl#L105),
```
Comment (sarahhunt) [inline doc/source/api/proposed_schema_changes.md]: switch to proto versions

File: doc/source/api/genotypephenotype.rst [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+contain a feature, phenotype, and/or evidence, which are combined as a
+logical AND to query the underlying datastore. Missing types are treated
+as a wildcard, returning all data.
+
+|image| http://yuml.me/edit/bf06b90a
+
+Response
+~~~~~~~~
+
+`Responses <https://github.com/ga4gh/schemas/blob/be171b00a5f164836dfd40ea5ae75ea56924d316/src/main/resources/avro/genotypephenotypemethods.avdl#L130>`__
```
Comment (sarahhunt) [inline doc/source/api/genotypephenotype.rst]: switch to proto versions

File: doc/source/api/genotypephenotype.rst [hunk 9]
```diff
@@ -0,0 +1,528 @@
+Summary
+=======
+
+This API endpoints allows users to search for genotype-phenotype
+associations in a GA4GH datastore. The user can search for associations
+by building queries composed of features, phenotypes, and/or evidence
+terms. The API is designed to accommodate search terms specified as
+either a string, external identifier, ontology identifier, or as an
```
Comment (david4096) [inline doc/source/api/genotypephenotype.rst]: Is this still the case? The query patterns seem to have drifted from the docs here.

File: doc/source/api/genotypephenotype.rst [hunk 10]
```diff
@@ -0,0 +1,528 @@
...
+`evidence <https://github.com/ga4gh/schemas/blob/be171b00a5f164836dfd40ea5ae75ea56924d316/src/main/resources/avro/genotypephenotypemethods.avdl#L111>`__.
+
+API
+---
+
+The G2P schemas define several endpoints broken into two entity searches
+and an association search.
+
+A feature or phenotype can potentially be represented in increasing
+specificity as either [a string, an ontology identifier, an external
```
Comment (david4096) [inline doc/source/api/genotypephenotype.rst]: Same as above. I believe the offering does not provide the entity or external identifier functionality.

File: doc/source/api/genotypephenotype.rst [hunk 11]
```diff
@@ -0,0 +1,528 @@
...
+
+The system will respond with phenotypes whose qualifiers that match that
+ontology 'is\_a'.
+
+Q: I have a disease name "inflammatory bowel disease".
+
+Create an PhenotypeQuery using description field.
+``{"description": "inflammatory bowel disease",...}`` The system
+responds with Phenotypes that match on OntologyTerm.description Note
+that you can wildcard description. ``{"description": ".*bowel.*",...}``
```
Comment (david4096) [inline doc/source/api/genotypephenotype.rst]: I don't believe that the regex option should be here.

**Comment by kozbo** [comment]
+1

File: src/main/proto/ga4gh/genotype_phenotype_service.proto [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    string description = 2;  /*regex*/
+
+    repeated ExternalIdentifier external_identifiers = 3; /* new field */
+}
+
+message SearchPhenotypesRequest {
+    // The `PhenotypeAssociationSet` to search. Mandatory
+    string phenotype_association_set_id = 1;
+
+    string id = 2;
```
Comment (david4096) [inline src/main/proto/ga4gh/genotype_phenotype_service.proto]: Why is the `id` included in the search request? If you have an ID, would you be searching for it?

File: src/main/proto/ga4gh/genotype_phenotype_service.proto [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+be returned.  Similarly, queries with OntologyTerms should perform
+the subclass closure.
+
+Each query can be made against a string, an array of external
+identifers (such as for gene or SNP ids), ontology term ids, or
+full feature/phenotype/evidence objects.
+*/
+message SearchGenotypePhenotypeRequest {
+  // The `PhenotypeAssociationSet` to search. Mandatory
+  string phenotype_association_set_id = 1;
```
Comment (david4096) [inline src/main/proto/ga4gh/genotype_phenotype_service.proto]: In other portions of the proto, a developer can read the comments here to interpret how requests are to be handled. Could we say more about how these fields are handled? Ideally, a developer can integrate based on reading the inline comments alone.

File: src/main/proto/ga4gh/genotype_phenotype_service.proto [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+    // The system should support regular expression matches
+    string description = 3;
+
+    // type
+    OntologyTerm type = 4;
+
+    // terms should be OR'd together. e.g. (severe OR abnormal)
+    repeated OntologyTerm qualifiers = 5;
+
+    // Age of onset
```
Comment (david4096) [inline src/main/proto/ga4gh/genotype_phenotype_service.proto]: "Only return results that match this" "Are greater than this" etc.

File: src/main/proto/ga4gh/genotype_phenotype_service.proto [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+
+    // new field
+    repeated ExternalIdentifier external_identifiers = 3;
+}
+
+message SearchPhenotypesRequest {
+    // The `PhenotypeAssociationSet` to search. Mandatory
+    string phenotype_association_set_id = 1;
+
+    // Phenotype ID
```
Comment (david4096) [inline src/main/proto/ga4gh/genotype_phenotype_service.proto]: I presume this will be removed if a get-by-phenotype ID endpoint is added?

File: src/main/proto/ga4gh/genotype_phenotype_service.proto [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+
+// Evidence for the phenotype association.
+message EvidenceQuery {
+    // ECO or OBI is recommended
+    OntologyTerm evidenceType = 1;
+
+    // regex
+    string description = 2;
+
+    // new field
```
Comment (david4096) [inline src/main/proto/ga4gh/genotype_phenotype_service.proto]: "Only match Evidence messages that have any of these external identifiers"

File: src/main/proto/ga4gh/genotype_phenotype_service.proto [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+  // `SearchPhenotypesRequest` as the post body and will return a JSON version
+  // of `SearchPhenotypeAssociationSetsResponse`.
+  rpc SearchPhenotype(SearchPhenotypesRequest)
+      returns (SearchPhenotypesResponse);
+
+  // Gets a list of genotype-phenotype assoications accessible through the API.
+  // `POST /genotypephenotype/search` must accept a JSON version of
+  // `SearchPhenotypesRequest` as the post body and will return a JSON version
+  // of `SearchPhenotypeAssociationSetsResponse`.
+  rpc GetPhenotypeAssociations(SearchGenotypePhenotypeRequest)
```
Comment (david4096) [inline src/main/proto/ga4gh/genotype_phenotype_service.proto]: The naming convention should be changed so that search is for search and get is for get.

File: src/main/proto/ga4gh/genotype_phenotype_service.proto (hunk 15 above)
Comment (bwalsh) [inline src/main/proto/ga4gh/genotype_phenotype_service.proto]: Yes, will add todo

File: src/main/proto/ga4gh/genotype_phenotype.proto [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+    // or severity.
+    // For example, severe: http://purl.obolibrary.org/obo/PATO_0000396
+    // or abnormal: http://purl.obolibrary.org/obo/PATO_0000460
+    repeated OntologyTerm qualifier = 3;
+
+    // HPO is recommended, for example, subclasses of
+    // http://purl.obolibrary.org/obo/HP_0011007
+    // TODO: also allow quantitative recording?
+    OntologyTerm age_of_onset = 4;
+
```
Comment (mbaudis) [inline src/main/proto/ga4gh/genotype_phenotype.proto]: The time specification in `PhenotypeInstance.age_of_onset`  is not well defined:

* elsewhere in the schema time attributes are defined in ISO8601
* the [HPO link](http://purl.obolibrary.org/obo/PATO_00004) does not resolve
* an optimal implementation would allow both the use of standard ISO time (where age would be of the subtype "duration") and ontology terms for age classes ("childhood onset" or such), as is also discussed in the [Phenopackets](https://github.com/phenopackets/phenopacket-format/wiki/Overview-of-Phenotype-Exchange-Format) draft (though also incomplete there).

In conclusion, this has either be switched to string, ISO; or a proper documentation of the use has to be provided.

Comment (bwalsh) [inline src/main/proto/ga4gh/genotype_phenotype.proto]: @mbaudis Thanks so much for the review.  Since the comment was written, the url has changed.  I'll update the comment.

Comment (mbaudis) [inline src/main/proto/ga4gh/genotype_phenotype.proto]: @bwalsh That removes the inconsistencies I mentioned (link, quantitative format example) & fixes the commit.

However, we still need a schema-wide way to represent ages in both quantitative (more important!) and qualitative ways. I see 2 options:

* separate attributes (`age...` + `age_classe...`)
* `Age` object with both ontology and ISO (and again API/user handled preference/parsing)

Unfortunately one cannot automatically derive age classes unambiguously, and also quantitative values are in every way superior. But then, lots of resources will only have qualitative values attached ...

Comment (mbaudis) [inline src/main/proto/ga4gh/genotype_phenotype.proto]: @bwalsh @kozbo @david4096 Please see https://github.com/ga4gh/schemas/pull/712.

Comment (bwalsh) [inline src/main/proto/ga4gh/genotype_phenotype.proto]: @mbaudis Thanks again.  We have two other schema changes queued up to work on post this merge.  I'll add this to that list.
